File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    time = 0
    i = 0
    finish = 0
    ing = dict()# 몇번째 : 간거리
    while finish < len(truck_weights):
        time += 1    
        
        for v in list(ing.keys()):
            ing[v] += 1
            if ing[v] >= bridge_length:
                del ing[v]
                finish += 1
        
        if i < len(truck_weights) and len(ing) < bridge_length and weight >= sum([truck_weights[x] for x in ing.keys()]) + truck_weights[i]:
            ing[i] = 0
            i += 1
            
        # if time < 12:
        #     print(f"time:{time}, i:{i}. img:{ing}, finish:{finish}")
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    class Truck {
        private int weight;
        private int location;
        
        public Truck(int w) {
            this.weight = w;
            this.location = 1;
        }
    }
    
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Deque<Truck> bridge = new ArrayDeque<>();
        
        int i = 0;
        int time = 0;
        while(i < truck_weights.length) {
            time++;
            if(bridge.isEmpty()) {
                Truck newTruck = new Truck(truck_weights[i]);
                bridge.addLast(newTruck);
                i++;
                continue;
            }
            int weightSum = 0;
            for(Truck t : bridge) {
                t.location++;
                weightSum += t.weight;
            }
            if(bridge.peekFirst().location == bridge_length) {
                bridge.pollFirst();
            }
            if(bridge.size() <= bridge_length && weightSum + truck_weights[i] <= weight) {
                Truck newTruck = new Truck(truck_weights[i]);
                bridge.addLast(newTruck);
                i++;
            }
        }
        
        return time + bridge_length;
    }
}